File: go/binary_heap/heap_up.go
package binary_heap

// HeapUp should move element up at index i in a binary min-heap
func HeapUp(heap []int, i int) {
	val := heap[i]

	for j := (i - 1) / 2; i > 0 && val < heap[j]; i, j = j, (j-1)/2 {
		heap[i] = heap[j]
	}

	heap[i] = val
}

File: java/src/binary_heap/HeapUp.java
package binary_heap;

public class HeapUp {
	// heapUp should move element up at index i in a binary min-heap
	public static void heapUp(int[] heap, int i) {
//		throw new UnsupportedOperationException("TODO");

		var val = heap[i];

		for (var j = (i-1)/2; i > 0 && heap[j] > val; i = j, j = (j-1)/2) {
			heap[i] = heap[j];
		}

		heap[i] = val;
	}
}
